From 0ddc05a16dfa19f25d4a6f1e00d34d2fb5a6f7f0 Mon Sep 17 00:00:00 2001
From: Pengpeng Cui <pengpengx.cui@intel.com>
Date: Thu, 23 Jul 2015 17:33:28 +0800
Subject: [PATCH 454/462] atomisp2: css2401 and
 2401_legacy-irci_master_20150722_1500

Merge CSS2.1 for css2401 and 2401_legacy
for irci_master_20150722_1500 firmware tag

Change-Id: I20707e3f953228627e69dcc11f7a69c360db3545
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-42071
Signed-off-by: Pengpeng Cui <pengpengx.cui@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/66253
Approver: WeiX Meng <weix.meng@intel.com>
Integrator: WeiX Meng <weix.meng@intel.com>
Build: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Maintainer: WeiX Meng <weix.meng@intel.com>
---
 .../css/css_2401_csi2p_system/isys_dma_global.h    |  2 --
 .../css/css_2401_csi2p_system/pixelgen_global.h    |  3 ---
 .../media/pci/atomisp2/css/css_api_version.h       |  6 ++++-
 .../camera/drivers/media/pci/atomisp2/css/ia_css.h |  2 +-
 .../drivers/media/pci/atomisp2/css/ia_css_env.h    |  3 ++-
 .../media/pci/atomisp2/css/ia_css_host_data.h      |  2 +-
 .../drivers/media/pci/atomisp2/css/ia_css_types.h  |  2 +-
 .../media/pci/atomisp2/css/ia_css_version_data.h   |  2 +-
 .../atomisp2/css/isp/kernels/hdr/ia_css_hdr.host.c |  2 +-
 .../atomisp2/css/isp/kernels/hdr/ia_css_hdr.host.h |  2 +-
 .../css/isp/kernels/hdr/ia_css_hdr_param.h         |  2 +-
 .../css/isp/kernels/hdr/ia_css_hdr_types.h         |  2 +-
 .../atomisp2/css/runtime/isys/src/virtual_isys.c   |  2 --
 .../camera/drivers/media/pci/atomisp2/css/sh_css.c | 12 +++++-----
 .../media/pci/atomisp2/css/sh_css_firmware.c       |  4 ++--
 .../media/pci/atomisp2/css/sh_css_host_data.c      |  1 +
 .../drivers/media/pci/atomisp2/css/sh_css_mipi.c   |  4 ----
 .../media/pci/atomisp2/css/sh_css_param_shading.c  |  8 +++++--
 .../drivers/media/pci/atomisp2/css/sh_css_params.c | 27 ++++++++++++++++++----
 19 files changed, 52 insertions(+), 36 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_2401_csi2p_system/isys_dma_global.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_2401_csi2p_system/isys_dma_global.h
index feb2a7a..190628b 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_2401_csi2p_system/isys_dma_global.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_2401_csi2p_system/isys_dma_global.h
@@ -28,7 +28,6 @@ more details.
  *
  * DMA Port.
  *
- * zhengjie.lu@intel.com:
  * The DMA port definition for the input system
  * 2401 DMA is the duplication of the DMA port
  * definition for the CSS system DMA. It is duplicated
@@ -52,7 +51,6 @@ struct isys2401_dma_port_cfg_s {
  *
  * DMA Device.
  *
- * zhengjie.lu@intel.com:
  * The DMA device definition for the input system
  * 2401 DMA is the duplicattion of the DMA device
  * definition for the CSS system DMA. It is duplicated
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_2401_csi2p_system/pixelgen_global.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_2401_csi2p_system/pixelgen_global.h
index 46daf75..148dccf 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_2401_csi2p_system/pixelgen_global.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_2401_csi2p_system/pixelgen_global.h
@@ -20,7 +20,6 @@ more details.
  * Pixel-generator. ("pixelgen_global.h")
  */
 /*
- * zhengjie.lu@intel.com:
  * Duplicates "sync_generator_cfg_t" in "input_system_global.h".
  */
 typedef struct sync_generator_cfg_s sync_generator_cfg_t;
@@ -41,7 +40,6 @@ typedef enum {
 } pixelgen_tpg_mode_t;
 
 /*
- * zhengjie.lu@intel.com:
  * "pixelgen_tpg_cfg_t" duplicates parts of
  * "tpg_cfg_t" in "input_system_global.h".
  */
@@ -76,7 +74,6 @@ struct pixelgen_tpg_cfg_s {
 };
 
 /*
- * zhengjie.lu@intel.com:
  * "pixelgen_prbs_cfg_t" duplicates parts of
  * prbs_cfg_t" in "input_system_global.h".
  */
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_api_version.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_api_version.h
index 1c68210..b32610d 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_api_version.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/css_api_version.h
@@ -34,7 +34,7 @@ The version string has four dot-separated numbers, read left to right:
 
 */
 
-#define CSS_API_VERSION_STRING	"2.1.20.6"
+#define CSS_API_VERSION_STRING	"2.1.20.7"
 
 /*
 Change log
@@ -428,6 +428,10 @@ v2.1.20.6 (2 changes parallel), Frame format override
 v2.1.20.6 (2 changes parallel), Output_system_in_res Information
 - Output_system_in_res_info field added to pipe_info struct 
 
+v2.1.20.7, Temprarily disable memory debug features for SVOS.
+- Temporary commented out the additions to allow SKC testing till root cause found
+- Changed files ia_css_env.h and sh_css.c.
+
 */
 
 #endif /*__CSS_API_VERSION_H*/
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css.h
index 53dd88b..99429c3 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css.h
@@ -1,4 +1,4 @@
-/* Release Version: irci_master_20150716_0647 */
+/* Release Version: irci_master_20150722_1500 */
 /*
 Support for Intel Camera Imaging ISP subsystem.
 Copyright (c) 2010 - 2015, Intel Corporation.
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_env.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_env.h
index a84ba1e..4511cf1 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_env.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_env.h
@@ -49,12 +49,13 @@ struct ia_css_cpu_mem_env {
 	void (*flush)(struct ia_css_acc_fw *fw);
 	/**< Flush function to flush the cache for given accelerator. */
 
+	#if !defined(__SVOS__)
 	/* a set of matching functions with additional debug params */
 	void * (*alloc_ex)(size_t bytes, bool zero_mem, const char *caller_func, int caller_line);
 	/**< same as alloc above, only with additional debug parameters */
 	void (*free_ex)(void *ptr, const char *caller_func, int caller_line);
 	/**< same as free above, only with additional debug parameters */
-
+	#endif
 };
 
 /** Environment with function pointers for allocation of memory for the CSS.
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_host_data.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_host_data.h
index 87d3aaa..ee1f774 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_host_data.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_host_data.h
@@ -1,4 +1,4 @@
-/* Release Version: irci_master_20150716_0647 */
+/* Release Version: irci_master_20150722_1500 */
 /*
 Support for Intel Camera Imaging ISP subsystem.
 Copyright (c) 2010 - 2015, Intel Corporation.
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_types.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_types.h
index 8ea1461..63d2cb7 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_types.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_types.h
@@ -1,4 +1,4 @@
-/* Release Version: irci_master_20150716_0647 */
+/* Release Version: irci_master_20150722_1500 */
 /*
 Support for Intel Camera Imaging ISP subsystem.
 Copyright (c) 2010 - 2015, Intel Corporation.
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_version_data.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_version_data.h
index cb9387f..0f0de4c 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_version_data.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/ia_css_version_data.h
@@ -22,7 +22,7 @@ more details.
 #define __IA_CSS_VERSION_DATA_H
 
 
-#define CSS_VERSION_STRING "REL:20150716_29.4_1058; API:2.1.20.6; GIT:irci_20150715_1500__d882b0#d882b03c94aedb0eba523ec9f136faa1b1611731; SDK:; USER:viedifw; "
+#define CSS_VERSION_STRING "REL:20150722_30.3_2047; API:2.1.20.7; GIT:scci_20150721_1800__cc540b#cc540b0aaea35d2861d813960bb98d557e40153f; SDK:; USER:viedifw; "
 
 
 #endif
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr.host.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr.host.c
index 4690cfd..301d9dd 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr.host.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr.host.c
@@ -1,4 +1,4 @@
-/* Release Version: irci_master_20150716_0647 */
+/* Release Version: irci_master_20150722_1500 */
 /*
 Support for Intel Camera Imaging ISP subsystem.
 Copyright (c) 2010 - 2015, Intel Corporation.
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr.host.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr.host.h
index b79b49a..ee3c19d 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr.host.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr.host.h
@@ -1,4 +1,4 @@
-/* Release Version: irci_master_20150716_0647 */
+/* Release Version: irci_master_20150722_1500 */
 /*
 Support for Intel Camera Imaging ISP subsystem.
 Copyright (c) 2010 - 2015, Intel Corporation.
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr_param.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr_param.h
index 463103c..0788198 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr_param.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr_param.h
@@ -1,4 +1,4 @@
-/* Release Version: irci_master_20150716_0647 */
+/* Release Version: irci_master_20150722_1500 */
 /*
 Support for Intel Camera Imaging ISP subsystem.
 Copyright (c) 2010 - 2015, Intel Corporation.
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr_types.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr_types.h
index 1be40db..8fbb3d2 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr_types.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/isp/kernels/hdr/ia_css_hdr_types.h
@@ -1,4 +1,4 @@
-/* Release Version: irci_master_20150716_0647 */
+/* Release Version: irci_master_20150722_1500 */
 /*
 Support for Intel Camera Imaging ISP subsystem.
 Copyright (c) 2010 - 2015, Intel Corporation.
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/runtime/isys/src/virtual_isys.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/runtime/isys/src/virtual_isys.c
index dd54809..dc1019b 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/runtime/isys/src/virtual_isys.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/runtime/isys/src/virtual_isys.c
@@ -753,7 +753,6 @@ static bool calculate_ibuf_ctrl_cfg(
 	cfg->ib_buffer.lines			= channel->ib_buffer.lines;
 
 	/*
-	 * zhengjie.lu@intel.com:
 	 * "dest_buf_cfg" should be part of the input system output
 	 * port configuration.
 	 *
@@ -779,7 +778,6 @@ static bool calculate_ibuf_ctrl_cfg(
 	}
 
 	/*
-	 * zhengjie.lu@intel.com:
 	 * "items_per_store" is hard coded as "1", which is ONLY valid
 	 * when the CSI-MIPI long packet is transferred.
 	 *
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css.c
index 8930eaf..887cc1d 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css.c
@@ -848,8 +848,6 @@ static bool sh_css_translate_stream_cfg_to_input_system_input_port_attr(
 		}
 
 		/*
-		 * zhengjie.lu@intel.com:
-		 *
 		 * TODO
 		 * - Make "color_cfg" as part of "ia_css_tpg_config".
 		 */
@@ -868,8 +866,6 @@ static bool sh_css_translate_stream_cfg_to_input_system_input_port_attr(
 		isys_stream_descr->tpg_port_attr.delta_cfg.v_delta = stream_cfg->source.tpg.y_delta;
 
 		/*
-		 * zhengjie.lu@intel.com:
-		 *
 		 * TODO
 		 * - Make "sync_gen_cfg" as part of "ia_css_tpg_config".
 		 */
@@ -887,8 +883,6 @@ static bool sh_css_translate_stream_cfg_to_input_system_input_port_attr(
 		isys_stream_descr->prbs_port_attr.seed1 = stream_cfg->source.prbs.seed1;
 
 		/*
-		 * zhengjie.lu@intel.com:
-		 *
 		 * TODO
 		 * - Make "sync_gen_cfg" as part of "ia_css_prbs_config".
 		 */
@@ -1722,8 +1716,10 @@ ia_css_load_firmware(const struct ia_css_env *env,
 	/* make sure we initialize my_css */
 	if ((my_css.malloc != env->cpu_mem_env.alloc) ||
 		(my_css.free != env->cpu_mem_env.free) ||
+		#if !defined(__SVOS__)
 		(my_css.malloc_ex != env->cpu_mem_env.alloc_ex) ||
 		(my_css.free_ex != env->cpu_mem_env.free_ex) ||
+		#endif
 		(my_css.flush != env->cpu_mem_env.flush)
 		)
 	{
@@ -1731,8 +1727,10 @@ ia_css_load_firmware(const struct ia_css_env *env,
 
 		my_css.malloc = env->cpu_mem_env.alloc;
 		my_css.free = env->cpu_mem_env.free;
+		#if !defined(__SVOS__)
 		my_css.malloc_ex = env->cpu_mem_env.alloc_ex;
 		my_css.free_ex = env->cpu_mem_env.free_ex;
+		#endif
 		my_css.flush = env->cpu_mem_env.flush;
 	}
 
@@ -1822,8 +1820,10 @@ ia_css_init(const struct ia_css_env *env,
 
 	malloc_func    = env->cpu_mem_env.alloc;
 	free_func      = env->cpu_mem_env.free;
+	#if !defined(__SVOS__)
 	malloc_func_ex = env->cpu_mem_env.alloc_ex;
 	free_func_ex   = env->cpu_mem_env.free_ex;
+	#endif
 	flush_func     = env->cpu_mem_env.flush;
 
 	pipe_global_init();
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_firmware.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_firmware.c
index b670712..260ce77 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_firmware.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_firmware.c
@@ -47,10 +47,10 @@ struct fw_param {
 /* Warning: same order as SH_CSS_BINARY_ID_* */
 static struct firmware_header *firmware_header;
 
-/* The string STR(irci_master_20150716_0647) is a place holder
+/* The string STR(irci_master_20150722_1500) is a place holder
  * which will be replaced with the actual RELEASE_VERSION
  * during package generation. Please do not modify  */
-static const char *release_version = STR(irci_master_20150716_0647);
+static const char *release_version = STR(irci_master_20150722_1500);
 
 #define MAX_FW_REL_VER_NAME	300
 static char FW_rel_ver_name[MAX_FW_REL_VER_NAME] = "---";
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_host_data.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_host_data.c
index 87c3bb3..56a6a91 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_host_data.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_host_data.c
@@ -36,6 +36,7 @@ void ia_css_host_data_free(struct ia_css_host_data *me)
 {
 	if (me) {
 		sh_css_free(me->address);
+		me->address = NULL;
 		sh_css_free(me);
 	}
 }
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_mipi.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_mipi.c
index 01d0711..6f6e7b2 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_mipi.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_mipi.c
@@ -305,8 +305,6 @@ calculate_mipi_buff_size(
 	enum ia_css_err err = IA_CSS_SUCCESS;
 
 	/**
-	 * zhengjie.lu@intel.com
-	 *
 	 * NOTE
 	 * - In the struct "ia_css_stream_config", there
 	 *   are two members: "input_config" and "isys_config".
@@ -322,8 +320,6 @@ calculate_mipi_buff_size(
 	/** end of NOTE */
 
 	/**
-	 * zhengjie.lu@intel.com
-	 *
 	 * NOTE
 	 * - The following code is derived from the
 	 *   existing code "ia_css_mipi_frame_calculate_size()".
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_param_shading.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_param_shading.c
index 757a509..4c967bc 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_param_shading.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_param_shading.c
@@ -354,8 +354,10 @@ ia_css_shading_table_alloc(
 		    sh_css_malloc(width * height * sizeof(*me->data[0]));
 		if (me->data[i] == NULL) {
 			unsigned int j;
-			for (j = 0; j < i; j++)
+			for (j = 0; j < i; j++) {
 				sh_css_free(me->data[j]);
+				me->data[j] = NULL;
+			}
 			sh_css_free(me);
 			return NULL;
 		}
@@ -379,8 +381,10 @@ ia_css_shading_table_free(struct ia_css_shading_table *table)
 	IA_CSS_ENTER("");
 
 	for (i = 0; i < IA_CSS_SC_NUM_COLORS; i++) {
-		if (table->data[i])
+		if (table->data[i]) {
 			sh_css_free(table->data[i]);
+			table->data[i] = NULL;
+		}
 	}
 	sh_css_free(table);
 
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_params.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_params.c
index 5322d51..3f0472e 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_params.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/css/sh_css_params.c
@@ -1953,10 +1953,14 @@ void ia_css_morph_table_free(
 
 
 	for (i = 0; i < IA_CSS_MORPH_TABLE_NUM_PLANES; i++) {
-		if (me->coordinates_x[i])
+		if (me->coordinates_x[i]) {
 			sh_css_free(me->coordinates_x[i]);
-		if (me->coordinates_y[i])
+			me->coordinates_x[i] = NULL;
+		}
+		if (me->coordinates_y[i]) {
 			sh_css_free(me->coordinates_y[i]);
+			me->coordinates_y[i] = NULL;
+		}
 	}
 
 	sh_css_free(me);
@@ -2252,8 +2256,11 @@ void
 ia_css_isp_3a_statistics_map_free(struct ia_css_isp_3a_statistics_map *me)
 {
 	if (me) {
-		if (me->data_allocated)
+		if (me->data_allocated) {
 			sh_css_free(me->data_ptr);
+			me->data_ptr = NULL;
+			me->data_allocated = false;
+		}
 		sh_css_free(me);
 	}
 }
@@ -3059,6 +3066,7 @@ ia_css_dvs_statistics_free_isp(enum dvs_statistics_type type,
 		if (DVS_STATISTICS == type || DVS2_STATISTICS == type) {
 			mmgr_free(me->p_dvs_statistics_isp->hor_proj);
 			mmgr_free(me->p_dvs_statistics_isp->ver_proj);
+			memset(me->p_dvs_statistics_isp, 0, sizeof(struct ia_css_isp_dvs_statistics));
 			sh_css_free(me->p_dvs_statistics_isp);
 		}
 	}
@@ -3799,8 +3807,10 @@ ia_css_stream_isp_parameters_uninit(struct ia_css_stream *stream)
 		free_map(&per_frame_params->ddr_ptrs);
 
 #if !defined(IS_ISP_2500_SYSTEM)
-	if (params->fpn_config.data)
+	if (params->fpn_config.data) {
 		sh_css_free(params->fpn_config.data);
+		params->fpn_config.data = NULL;
+	}
 #endif
 
 #if defined(IS_ISP_2500_SYSTEM)
@@ -3822,7 +3832,8 @@ ia_css_stream_isp_parameters_uninit(struct ia_css_stream *stream)
 	sh_css_free(params);
 	if (per_frame_params)
 		sh_css_free(per_frame_params);
-
+	stream->isp_params_configs = NULL;
+	stream->per_frame_isp_params_configs = NULL;
 	IA_CSS_LEAVE_PRIVATE("void");
 }
 
@@ -5269,6 +5280,7 @@ ia_css_3a_statistics_free(struct ia_css_3a_statistics *me)
 	if (me) {
 		sh_css_free(me->rgby_data);
 		sh_css_free(me->data);
+		memset(me, 0, sizeof(struct ia_css_3a_statistics));
 		sh_css_free(me);
 	}
 }
@@ -5308,6 +5320,7 @@ ia_css_dvs_statistics_free(struct ia_css_dvs_statistics *me)
 	if (me) {
 		sh_css_free(me->hor_proj);
 		sh_css_free(me->ver_proj);
+		memset(me, 0, sizeof(struct ia_css_dvs_statistics));
 		sh_css_free(me);
 	}
 }
@@ -5349,6 +5362,7 @@ ia_css_dvs_coefficients_free(struct ia_css_dvs_coefficients *me)
 	if (me) {
 		sh_css_free(me->hor_coefs);
 		sh_css_free(me->ver_coefs);
+		memset(me, 0, sizeof(struct ia_css_dvs_coefficients));
 		sh_css_free(me);
 	}
 }
@@ -5425,6 +5439,7 @@ ia_css_dvs2_statistics_free(struct ia_css_dvs2_statistics *me)
 		sh_css_free(me->ver_prod.odd_imag);
 		sh_css_free(me->ver_prod.even_real);
 		sh_css_free(me->ver_prod.even_imag);
+		memset(me, 0, sizeof(struct ia_css_dvs2_statistics));
 		sh_css_free(me);
 	}
 }
@@ -5501,6 +5516,7 @@ ia_css_dvs2_coefficients_free(struct ia_css_dvs2_coefficients *me)
 		sh_css_free(me->ver_coefs.odd_imag);
 		sh_css_free(me->ver_coefs.even_real);
 		sh_css_free(me->ver_coefs.even_imag);
+		memset(me, 0, sizeof(struct ia_css_dvs2_coefficients));
 		sh_css_free(me);
 	}
 }
@@ -5563,6 +5579,7 @@ ia_css_dvs2_6axis_config_free(struct ia_css_dvs_6axis_config *dvs_6axis_config)
 		sh_css_free(dvs_6axis_config->ycoords_y);
 		sh_css_free(dvs_6axis_config->xcoords_uv);
 		sh_css_free(dvs_6axis_config->ycoords_uv);
+		memset(dvs_6axis_config, 0, sizeof(struct ia_css_dvs_6axis_config));
 		sh_css_free(dvs_6axis_config);
 	}
 }
-- 
1.9.1

